### ISSUE TITLE
java.lang.LinkageError when using Datadog Java agent alongside Opentelemetry Java agent

### ISSUE THREAD

**Issue by PhilHardwick** [DataDog/dd-trace-java#4435]
This issue is partly to make you aware and partly for anyone else who may experience this.

As context to this, we use the Datadog Java agent for metrics collection and Otel Java agent for trace instrumentation.

When we upgraded our Datadog java agent to any version beyond 0.108.0 we got the below error (with Otel Java agent v1.20.1) whenever a service consumed its first message from Kafka. This caused Context to leak, the message processing span never to end or be reported and all future messaging processing to attach to the same trace.

```
java.lang.IllegalStateException: This error handler cannot process 'java.lang.IllegalStateException's; no record information is available
	at org.springframework.kafka.listener.DefaultErrorHandler.handleOtherException(DefaultErrorHandler.java:155)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.handleConsumerException(KafkaMessageListenerContainer.java:1787)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1296)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalStateException: java.lang.LinkageError: loader 'app' attempted duplicate class definition for io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.AutoValue_State. (io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.AutoValue_State is in unnamed module of loader 'app')
	at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$UsingUnsafeInjection.defineClass(ClassInjector.java:1045)
	at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection.injectRaw(ClassInjector.java:284)
	at io.opentelemetry.javaagent.tooling.HelperInjector$HelperClassInjector.inject(HelperInjector.java:370)
	at io.opentelemetry.javaagent.tooling.HelperInjector.loadHelperClass(HelperInjector.java:357)
	at io.opentelemetry.javaagent.bootstrap.InjectedClassHelper.loadHelperClass(InjectedClassHelper.java:55)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.InstrumentedRecordInterceptor.start(InstrumentedRecordInterceptor.java:53)
	at io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.InstrumentedRecordInterceptor.intercept(InstrumentedRecordInterceptor.java:43)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.checkEarlyIntercept(KafkaMessageListenerContainer.java:2457)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeWithRecords(KafkaMessageListenerContainer.java:2422)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeRecordListener(KafkaMessageListenerContainer.java:2305)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeListener(KafkaMessageListenerContainer.java:1979)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeIfHaveRecords(KafkaMessageListenerContainer.java:1364)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.pollAndInvoke(KafkaMessageListenerContainer.java:1355)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1247)
	... 3 common frames omitted
Caused by: java.lang.LinkageError: loader 'app' attempted duplicate class definition for io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.AutoValue_State. (io.opentelemetry.javaagent.shaded.instrumentation.spring.kafka.v2_7.AutoValue_State is in unnamed module of loader 'app')
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
	at java.base/java.lang.ClassLoader$ByteBuddyAccessor$V1.defineClass(Unknown Source)
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor24.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$UsingUnsafeInjection.defineClass(ClassInjector.java:1041)
	... 17 common frames omitted
```

The fix/workaround for this was to set `-Ddd.trace.classes.exclude=io.opentelemetry.*`. It seems like Datadog is processing the Otel classes and therefore loading them. When the Otel Java agent tries to load the class it finds one already in the classloader.

Is this expected? This didn't happen in version 0.108.0 and before.

**Comment by mcculls** [comment]
This will be fixed by https://github.com/DataDog/dd-trace-java/pull/4964 which will soon be available in 1.11.0

**Comment by dmitry-cherkas** [comment]
This is indeed not reproducible with dd agent `1.11.2`
